File: architecture-skills/design-patterns/examples/builder/builder.cs
// Builder pattern in C#.
//
// Note: C# object initializers and records with `init` setters cover simple
// cases; use a builder for multi-stage construction with validation.

using System;
using System.Collections.Generic;

// Immutable product.
public sealed record HttpRequest(
    string Method,
    string Url,
    IReadOnlyDictionary<string, string> Headers,
    string? Body,
    double TimeoutSeconds);

public sealed class HttpRequestBuilder
{
    private readonly string _method;
    private readonly string _url;
    private readonly Dictionary<string, string> _headers = new();
    private string? _body;
    private double _timeoutSeconds = 30.0;

    public HttpRequestBuilder(string method, string url)
    {
        _method = method.ToUpperInvariant();
        _url = url;
    }

    public HttpRequestBuilder Header(string name, string value)
    {
        _headers[name] = value;
        return this;
    }

    public HttpRequestBuilder Body(string body)
    {
        _body = body;
        return this;
    }

    public HttpRequestBuilder Timeout(double seconds)
    {
        _timeoutSeconds = seconds;
        return this;
    }

    public HttpRequest Build()
    {
        if (_body is not null && _method is "GET" or "HEAD")
        {
            throw new InvalidOperationException($"{_method} requests cannot have a body");
        }
        return new HttpRequest(_method, _url, new Dictionary<string, string>(_headers), _body, _timeoutSeconds);
    }
}

public static class Program
{
    public static void Main()
    {
        HttpRequest request = new HttpRequestBuilder("POST", "https://api.example.com/users")
            .Header("Content-Type", "application/json")
            .Header("Authorization", "Bearer token123")
            .Body("{\"name\": \"Ada\"}")
            .Timeout(5.0)
            .Build();

        Console.WriteLine($"{request.Method} {request.Url} (timeout {request.TimeoutSeconds}s)");
        foreach (var (name, value) in request.Headers)
        {
            Console.WriteLine($"{name}: {value}");
        }
        Console.WriteLine(request.Body);
    }
}

File: architecture-skills/design-patterns/examples/builder/builder.php
<?php
// Builder pattern in PHP.
//
// Note: named arguments (PHP 8+) cover simple cases; use a builder for
// multi-stage construction with validation.

declare(strict_types=1);

/** Immutable product. */
final class HttpRequest
{
    /** @param array<string, string> $headers */
    public function __construct(
        public readonly string $method,
        public readonly string $url,
        public readonly array $headers,
        public readonly ?string $body,
        public readonly float $timeoutSeconds,
    ) {
    }
}

final class HttpRequestBuilder
{
    /** @var array<string, string> */
    private array $headers = [];
    private ?string $body = null;
    private float $timeoutSeconds = 30.0;

    public function __construct(
        private readonly string $method,
        private readonly string $url,
    ) {
    }

    public function header(string $name, string $value): self
    {
        $this->headers[$name] = $value;
        return $this;
    }

    public function body(string $body): self
    {
        $this->body = $body;
        return $this;
    }

    public function timeout(float $seconds): self
    {
        $this->timeoutSeconds = $seconds;
        return $this;
    }

    public function build(): HttpRequest
    {
        $method = strtoupper($this->method);
        if ($this->body !== null && in_array($method, ['GET', 'HEAD'], true)) {
            throw new InvalidArgumentException("$method requests cannot have a body");
        }
        return new HttpRequest($method, $this->url, $this->headers, $this->body, $this->timeoutSeconds);
    }
}

$request = (new HttpRequestBuilder('POST', 'https://api.example.com/users'))
    ->header('Content-Type', 'application/json')
    ->header('Authorization', 'Bearer token123')
    ->body('{"name": "Ada"}')
    ->timeout(5.0)
    ->build();

var_dump($request);
